### PR TITLE
[vcpkg-tool-meson] Refactor cmd line args generation

### PR DESCRIPTION
**Pull request by Neumann-A**
introduce `vcpkg_generate_meson_cmd_args` to generate the command line args vcpkg calls meson with. This needed to be factored out since it allows calling the meson-py backend of setuptools with the same arguments as calling meson manually within vcpkg. 

additionally:
- fixed `CMAKE_BUILD_TYPE` (var `buildconfig` does not exist)
- add `ADDITIONAL_PROPERTIES` option
- removed a lot of stuff being added to PATH unnecessarily 
- added option `--pkgconfig.relocatable` (probably not necessary but we want this any way.)
- added option `--pkg-config-path` with correct paths (allows removing the env variables for it)

cc @dg0yt

### CODE REVIEW

**Comment by dg0yt** [comment]
Is this still `vcpkg-tool-meson`, or is it time to introduce `vcpkg-meson`?

**Comment by Neumann-A** [comment]
> Is this still vcpkg-tool-meson, or is it time to introduce vcpkg-meson?

The chain is `python->setuptools->meson-py->meson`. So in the end `meson` is still going to be called. (Just not directly by vcpkg)
However, I am thinking to add another wrapper port around meson so that I can also depend on `pkgconf` without creating a cycle on windows. The problem is that to do that without touching all the ports depending on `vcpkg-tool-meson` would require making the wrapper itself `vcpkg-tool-meson` and move everything from it into another port. This way `vcpkg-tool-meson` could depend on `pkgconf` and the new port and `pkgconf` only depends on the new port. This way the dependency on pkgconf from msys could get removed.

**Comment by Neumann-A** [comment]
Hmm some issue with `kf5kio` probably having a badly escaped generator expression in the build.ninja file. Impossible to debug without having that file.

**Comment by dg0yt** [comment]
> Hmm some issue with `kf5kio` probably having a badly escaped generator expression in the build.ninja file. Impossible to debug without having that file.

https://github.com/microsoft/vcpkg/pull/37196#issuecomment-2133284249

**Comment by Neumann-A** [comment]
That explains why I couldn't reproduce it locally since krb5 is not declared as a dependency.

**Review by BillyONeal — COMMENTED** [review]
Seems mostly reasonable to me. 👍

File: ports/vcpkg-tool-meson/vcpkg_configure_meson.cmake [hunk 1]
```diff
@@ -327,114 +320,148 @@ function(vcpkg_configure_meson)
...
-    # Allow overrides / additional configuration variables from triplets
-    if(DEFINED VCPKG_MESON_CONFIGURE_OPTIONS)
-        vcpkg_list(APPEND arg_OPTIONS ${VCPKG_MESON_CONFIGURE_OPTIONS})
+    file(REMOVE_RECURSE "${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-rel")
+    file(REMOVE_RECURSE "${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-dbg")
+
+    vcpkg_find_acquire_program(MESON)
+
+    get_filename_component(CMAKE_PATH "${CMAKE_COMMAND}" DIRECTORY)
+    vcpkg_add_to_path("${CMAKE_PATH}") # Make CMake invokeable for Meson
```
[the remark this answers is not in the document]
Comment (Neumann-A) [reply ports/vcpkg-tool-meson/vcpkg_configure_meson.cmake:420]: Was to make sure that the correct CMake is used (VS vs vcpkg). However it seriously breaks stuff if CMake is in `/bin` with other stuff. 

Simply needs a better function to control program injection into PATH.

File: ports/vcpkg-tool-meson/vcpkg_configure_meson.cmake [hunk 2]
```diff
@@ -297,28 +297,21 @@ function(z_vcpkg_meson_setup_variables config_type)
...
-    vcpkg_list(APPEND arg_ADDITIONAL_BINARIES ${arg_ADDITIONAL_NATIVE_BINARIES} ${arg_ADDITIONAL_CROSS_BINARIES})
-    vcpkg_list(REMOVE_DUPLICATES arg_ADDITIONAL_BINARIES)
     vcpkg_list(JOIN arg_ADDITIONAL_BINARIES "\n" MESON_ADDITIONAL_BINARIES)
+    vcpkg_list(JOIN arg_ADDITIONAL_PROPERTIES "\n" MESON_ADDITIONAL_PROPERTIES)
 
-    file(REMOVE_RECURSE "${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-rel")
-    file(REMOVE_RECURSE "${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-dbg")
+    set(buildtype "${arg_CONFIG}")
 
     if(NOT VCPKG_CHAINLOAD_TOOLCHAIN_FILE)
```
[the remark this answers is not in the document]
Comment (Neumann-A) [reply ports/vcpkg-tool-meson/vcpkg_configure_meson.cmake:316]: `z_vcpkg_get_cmake_vars` 
 uses a cache variable to not unnecessary be rerun https://github.com/microsoft/vcpkg/blob/eb0f108ebd674c6ed79acb1c2e123208c416af0d/scripts/cmake/z_vcpkg_get_cmake_vars.cmake#L15

File: ports/vcpkg-tool-meson/vcpkg_configure_meson.cmake [hunk 3]
```diff
@@ -297,28 +297,21 @@ function(z_vcpkg_meson_setup_variables config_type)
...
-    endif()
-
-    vcpkg_list(APPEND arg_ADDITIONAL_BINARIES ${arg_ADDITIONAL_NATIVE_BINARIES} ${arg_ADDITIONAL_CROSS_BINARIES})
-    vcpkg_list(REMOVE_DUPLICATES arg_ADDITIONAL_BINARIES)
     vcpkg_list(JOIN arg_ADDITIONAL_BINARIES "\n" MESON_ADDITIONAL_BINARIES)
+    vcpkg_list(JOIN arg_ADDITIONAL_PROPERTIES "\n" MESON_ADDITIONAL_PROPERTIES)
 
-    file(REMOVE_RECURSE "${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-rel")
-    file(REMOVE_RECURSE "${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-dbg")
+    set(buildtype "${arg_CONFIG}")
```
[the remark this answers is not in the document]
Comment (Neumann-A) [reply ports/vcpkg-tool-meson/vcpkg_configure_meson.cmake:314]: I just did it to avoid a lot of code changes which might introduce errors.

**Review by BillyONeal — APPROVED** [review]
OK not worth blocking over a naming nitpick. Thanks for the responses!

**Comment by BillyONeal** [comment]
Thanks for the extract function :)

**Pull request merged**